pipeline/authn/authenticator_oauth2_introspection.go

```diff
@@ -96,6 +98,35 @@ type AuthenticatorOAuth2IntrospectionResult struct {
 	TokenUse  string                 `json:"token_use"`
 }
 
+func (a *Audience) UnmarshalJSON(b []byte) error {
+	var errUnsupportedType = errors.New("Unsupported aud type, only string or []string are allowed")
+
+	var jsonObject interface{}
```

Comment:
Since it could also be `[]string` I believe it might make sense to use gjson to assert the type and then decode into the right struct. WDYT?

Reply:
I don't think this is necessary, as I'm already checking for `[]string` below. No need to use gjson here.

Reply:
:thinking: I only see cases for `[]interface{}` and `string`

Reply:
As far as I see the code checks for `string` and `[]interface`. AFAIK it's possible that the JSON decoder also converts to `[]string`.

Reply:
@zepatrik @aeneasr Correct, the JSON decoder supports `[]string`. In this case we don't know whether it's `[]string` or `string` ahead of time, so we can't just decode into `[]string`. AFAIK this is the cleanest solution:
1. Unmarshal into generic `interface{}`
2. Check if type is `string`
3. Check if type is generic `[]interface{}` and build a `[]string`.

I tried to check for `[]string` instead of `[]interface{}`, it doesn't work. My Go skills are a little bit rusty as it has been a few years, although in this case I'm pretty confident it's a good solution. Do you have a better proposal? What don't you like about this solution?

Reply:
I think the problem is that there are four possible cases:

1. `string`
1. `[]string`
1. `[]interface{}`

because `[]string` != `[]interface{}` - you need type conversion :)

So another solution might look like this:

```
var auds []string
if aud := gjson.Get(json, "aud"); aud.IsArray() {
  aud.ForEach(func(_, result) {
    auds = append(auds, result)
  })
} else if len(aud.Str) > 0 {
  auds = append(aud.Str)
} else {
  // error or something else
}
```

Reply:
Just for my understanding, can you provide an input where the current conversion fails to produce a valid output? Still don't see the problem.

Reply:
@aeneasr 
> because `[]string` != `[]interface{}` just catching `[]interface{}` is not enough - you need type conversion :)

https://github.com/ory/oathkeeper/pull/822/files#diff-5c8061290da17d352a4b888e5270cdd071ea584cd4751011132d8cccd69594f0R117

This looks like type conversion :) I'm just checking if it's a generic array, then cast all members to a string and fail if a member is not a string. The solution works without using external libraries like gjson. If you prefer to use gjson, feel free to edit the PR. If not, please consider merging it soon.

Reply:
Go does not have subtyping rules, so `[]interface{}{} != []string{}`. What you liked is not a type conversion, but type assertion. Go will look whether the value is of the specified type, and `ok` is true if it is.
https://play.golang.org/p/DCbHNxY0JmJ gives you a type error although one is the sub type of the other.

But regardless, `json.Unmarshal` does not seem to give you a `[]string`: https://play.golang.org/p/MX9XbCzjCth so this code should be fine.

Reply:
@zepatrik @aeneasr Can you please merge this?

Reply:
Oh, missed @zepatrik comment! Thank you :)